Userland/Libraries/LibCore/SharedCircularQueue.h

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2022, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Assertions.h>
+#include <AK/Atomic.h>
+#include <AK/BuiltinWrappers.h>
+#include <AK/Error.h>
+#include <AK/Format.h>
+#include <AK/Function.h>
+#include <AK/NonnullRefPtr.h>
+#include <AK/Platform.h>
+#include <AK/RefCounted.h>
+#include <AK/RefPtr.h>
+#include <AK/String.h>
+#include <AK/Types.h>
+#include <AK/Variant.h>
+#include <AK/Weakable.h>
+#include <LibCore/AnonymousBuffer.h>
+#include <LibCore/System.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sched.h>
+#include <sys/mman.h>
+
+namespace Core {
+
+// A circular lock-free queue (or a buffer) with a single producer,
+// residing in shared memory and designed to be accessible to multiple processes.
+// This implementation makes use of the fact that any producer-related code can be sure that
+// it's the only producer-related code that is running, which simplifies a bunch of the synchronization code.
+// The exclusivity and liveliness for critical sections in this class is proven to be correct
+// under the assumption of correct synchronization primitives, i.e. atomics.
+// In many circumstances, this is enough for cross-process queues.
+// This class is designed to be transferred over IPC and mmap()ed into multiple processes' memory.
+// It is a synthetic pointer to the actual shared memory, which is abstracted away from the user.
+// clang-format off
+template<typename T, size_t Size = 32>
+// Size must be a power of two, which speeds up the modulus operations for indexing.
+requires(popcount(Size) == 1)
+class SharedSingleProducerCircularQueue final {
+
+public:
+    using ValueType = T;
+
+    enum class QueueStatus : u8 {
+        Invalid = 0,
+        Full,
+        Empty,
+    };
+
+    SharedSingleProducerCircularQueue() = default;
+    // We need copy and move constructors because at the IPC boundary we can't throw around smart pointers.
```

Comment:
I'm having trouble following what this comment is suggesting.
Why/how would a `SharedSingleProducerCircularQueue` ever be passed over IPC?

Wouldn't you just need access to the `fd` and then create it on the other side via `SharedSingleProducerCircularQueue::try_create(..)`.

(Still have only looked at this early commit, not the follow ups yet.)

Reply:
3923e0f5a13382cce2f17c6c83cabe24cd5928ba seems to confirm my understanding. So I'm confused why this need to be copyable / movable then, and why the inner queue is a `RefPtr<..>`?

Reply:
I discussed this with Ali and we settled on this solution: The outer class is mostly a wrapper around a RefPtr to the inner container that "owns" the shared memory. This is necessary specifically because of IPC, as the IPC encoding, decoding and transferring methods expects classes to be copyable and movable. So when we copy this outer class, we simply increment the refcount on the inner container. And yes, over IPC we only send the fd. The entire reference handling prodedure happens inside one process, it doesn't interact with the shared memory itself.

Looking at the comment again, it's from an old revision that didn't have this desing. I think I'll just remove it.